src/libsync/owncloudpropagator.cpp

```diff
@@ -448,6 +448,44 @@ void OwncloudPropagator::start(SyncFileItemVector &&items)
             items.end());
     }
 
+    // process each item that is new and is a directory and make sure every parent in its tree has the instruction NEW instead of REMOVE
+    for (const auto &item : items) {
+        if (item->_instruction == CSYNC_INSTRUCTION_NEW && item->_direction == SyncFileItem::Up && item->isDirectory()) {
+            const auto folderPathSplit = item->_file.split(QLatin1String("/"), Qt::SkipEmptyParts);
+            // #1 get root folder name for the current item that we need to reupload
+            const auto itemRootFolderName = folderPathSplit.first();
+            // #2 iterate backwords (for optimization) and find the root folder by name
+            const auto itemRootFolderReverseIt = std::find_if(std::rbegin(items), std::rend(items), [&itemRootFolderName](auto &currentItem) {
+                return currentItem->_instruction == CSYNC_INSTRUCTION_REMOVE && currentItem->_direction == SyncFileItem::Down && currentItem->_file == itemRootFolderName;
+            });
+
+            if (itemRootFolderReverseIt != std::rend(items)) {
+
+                // #3 convert reverse iterator to normal iterator
+                const auto itemFolderIt = (itemRootFolderReverseIt + 1).base();
+
+                // #4 if the root folder is set to be removed, then we will need to fix it by reupload every folder in the tree, including the root
+                if (itemFolderIt != std::end(items) && (*itemFolderIt)->_instruction == CSYNC_INSTRUCTION_REMOVE) {
+                    qCWarning(lcPropagator) << "WARNING:  Job within a removed directory?  This should not happen! But, we are going to reupload the entire folder structure."
+                                            << item->_file << item->_instruction;
+
+                    auto nextFolderInTreeIt = itemFolderIt;
+                    do {
+                        // #5 Iterate forward from the root until the item that we wanted to reupload, and make sure every folder is set to reupload too
+                        if ((*nextFolderInTreeIt)->isDirectory()
+                            && (*nextFolderInTreeIt)->_instruction == CSYNC_INSTRUCTION_REMOVE
+                            && (*nextFolderInTreeIt)->_direction == SyncFileItem::Down
+                            && item->_file.startsWith((*nextFolderInTreeIt)->_file)) {
+                                (*nextFolderInTreeIt)->_instruction = CSYNC_INSTRUCTION_NEW;
```

Comment:
I now that conflict on folders is apparently not working (at least according to the comments within the code) but what you want may be `CSYNC_INSTRUCTION_NONE` or `CSYNC_INSTRUCTION_IGNORE` to delay deleting the folder once every item inside it have been propagated
putting NEW might trigger an upload of the folder and this is not necessarily what we want in this case
we could end up with a fight between users wanting to upload files within a folder and another user trying to delete it
maybe have a look at how you could generate conflicts NEW/DELETE for folders

Reply:
@mgallien Then we have a problem. Let's say folder "A" is gone from the server. Then, let's say, our user had problems uploading some new files into "A/B/C/c1". Then, let's say we don't want to remove folder "c1" with files that the user has not been able to upload yet, but, do not have any copy of those files. Then, we would still want to upload a folder "c1" into the server's "A/B/C/" with the same structure as it was before someone deleted "A" with all subfolders from the server.

If we just try to upload a folder "c1" to the server, using the old location, will that then re-create a folder structure on the server? Will "c1" be uploaded into "A/B/C/" ? If this will happen, then, what are we changing with using `CSYNC_INSTRUCTION_NONE` or `CSYNC_INSTRUCTION_IGNORE`? Otherwise, if the folder structure will not be recreated by the server when using `CSYNC_INSTRUCTION_NONE` or `CSYNC_INSTRUCTION_IGNORE`, then, we will not be able to upload the folder "c1" back to the server. Yes, files will probably not get removed, but, the user's client will stop synchronizing until manual action is taken by moving files elsewhere, etc. When setting `CSYNC_INSTRUCTION_NEW` we will just silently upload only those files that were just brought by the user, and will not upload those files that were on the server but were deleted. At least, this is how it works now. If I have added 5 files, waited until they got uploaded to the server, exited the desktop client, removed the entire "A" folder from the server, added one new file to a folder "c1", restarted the desktop client, and, had only one file uploaded with entire directory structure recovered on the server. So, it happens silently from the user's perspective with no need to resolve conflicts or lose files.

Having said that, I haven't tried using `CSYNC_INSTRUCTION_NONE` or `CSYNC_INSTRUCTION_IGNORE` instead of `CSYNC_INSTRUCTION_NEW` and see how it goes.

Reply:
@mgallien I have tried your suggestion of using `CSYNC_INSTRUCTION_NONE` or `CSYNC_INSTRUCTION_IGNORE`. Unfortunately, they both result in a folder still being deleted during the second sync, and files are lost.